### DISCUSSION TITLE
Use matplot++ on Linux server

### DISCUSSION THREAD

**Discussion by avalonhse** [alandefreitas/matplotplusplus#43] [Q&A]
I am interested on using Matplot++ to generate plot and save them as images on server. 

The code runs well and generate the image as I want. However, every time running matplot++, it tries to open the gnuplot and fail, which generates the error message:

> QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-hoang'
> qt.qpa.screen: QXcbConnection: Could not connect to display :0
> Could not connect to any X display. 

Is there anyway to tell matplot++ do not open the gnu plot? It will make the library a perfect choice for many data science tasks.

**Comment by alandefreitas** [comment]
You can tell Gnuplot not to open a window.  The error you're getting is from the Qt terminal, so you should be fine.
But you shouldn't tell matplot++ not to open Gnuplot at all because you would also need Gnuplot to save the images.

To not open the window, create the figure in quiet mode:

```cpp
auto f = figure(true);
```

plot whatever you want, and then save:

```cpp
f->save("image_name.svg");
```

If you really need to avoid Gnuplot, even in this non-interactive mode, then the only solution would be to write a completely non-interactive backend for images. Probably something like AGG for vector graphics. 

It's also possible to adapt the OpenGL backend to save its buffer as JPG or even SVG, but I think it probably doesn't adapt well to your use case. I don't have a lot of experience with OpenGL. I know you can save the buffer but I don't if it's possible to completely bypass the interactive graphics like that.

See this [section](README.md#backends)

**Reply by Inferno171** [reply]
can you help to use opengl as backend in this lib ?

Thanks!!

**Reply by xakod** [reply]
Link is outdated : (

**Comment by alandefreitas** [comment]
Another option is to permanently set the output in the backend:

```cpp
auto f = figure(true);
f->backend()->output("file.svg");
```

Now, whenever you call:

```cpp
f->draw();
```

it will save an image in `file.svg`.

**Comment by avalonhse** [comment]
Great!! Thank you. Setting figure(true) is exactly what I want. Now I can call the executable file on a Jupiter notebook server for data science tasks easily.

The option OpenGL backend attracts me as well because my original interest with matplot++ is scientific applications on iOS devices. Currently I use Core-plot for iOS graphs but matplotlib is much less verbose for scientific applications. matplotlib-cpp needs python so it is not a choice in iOS.

Gnuplot seems not to work on iOS as well but there is OpenGL ES on iPhone. I am not sure does it work but if it did, matplot++ will be a great tool to push a lot scientific applications into practical use on phones.

**Reply by gddavis21** [reply]
I'm interested in matplot++ on iOS as well---have you gotten this to work?

**Comment by alandefreitas** [comment]
I'm glad it could solve your problem. A Jupiter notebook seems interesting. Let me know if there's anything public in that sense.

Yes. You can also use OpenGL in Qt applications, embed it in an ImGui style, or compile it to WebGL with Emscripten. I'm slowly implementing this backend as I need its plots for other applications. Line plots are fine. The others might take a while.

**Reply by Inferno171** [reply]
I want to use this library for plotting wxGrid's data (wxWidgets) but as Gnuplot requires terminal excess and Opengl doesn't so can you help me to use opengl as backend for plotting data?

**Reply by alandefreitas** [reply]
The current OpenGL backend is still experimental. It only works for a few types of plots. 
You can try this backend by building from source with the -BUILD_EXPERIMENTAL_OPENGL_BACKEND=ON option.

I'm currently working on a complete OpenGL backend but this is still going to take a few months.
It's probably going to become the default backend when done.

After that, creating backends for Qt, Wx, xeus-cling and even wasm should be relatively trivial.

**Reply by Inferno171** [reply]
I can't figure out where to put BUILD_EXPERIMENTAL_OPENGL_BACKEND=ON
To build with opengl . Can you please tell me a liitle bit more info about that.?
It would be verry helpfull!!!

**Reply by Inferno171** [reply]
And building from source cmake can't include all source file in xcode . Do you have fix for that??

**Reply by alandefreitas** [reply]
When [building from source](https://alandefreitas.github.io/matplotplusplus/integration/install/build-from-source/build-and-install/), replace your cmake command with:

```console
cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-O2" -DBUILD_EXPERIMENTAL_OPENGL_BACKEND=ON
```

Look at the proper commands for your operating system. You will need *some* experience with cmake to make this work though. You might need some extra [dependencies](https://alandefreitas.github.io/matplotplusplus/integration/install/build-from-source/dependencies/) too, like opengl, of course. The opengl backend is also **very** embryonic. More like a proof of concept.

**Reply by Inferno171** [reply]
Thanks 😊

**Comment by laoshaw** [comment]
what does matplot++ bring comparing to directly use gnuplot with c++?
